misc/multistream-select/src/dialer_select.rs

```diff
@@ -268,6 +450,182 @@ where
+                                    self.state = SimOpenState::SendRole {
+                                        io,
+                                        local_role: Role::Responder,
+                                    };
+                                }
+                            }
+                        }
+                        _ => return Poll::Ready(Err(ProtocolError::InvalidMessage.into())),
+                    }
+                }
+                SimOpenState::SendRole { mut io, local_role } => {
+                    match Pin::new(&mut io).poll_ready(cx)? {
+                        Poll::Ready(()) => {}
+                        Poll::Pending => {
+                            self.state = SimOpenState::SendRole { io, local_role };
+                            return Poll::Pending;
+                        }
+                    }
+
+                    let msg = match local_role {
+                        Role::Initiator => Message::Initiator,
+                        Role::Responder => Message::Responder,
+                    };
+
+                    if let Err(err) = Pin::new(&mut io).start_send(msg) {
+                        return Poll::Ready(Err(From::from(err)));
+                    }
+
+                    self.state = SimOpenState::FlushRole { io, local_role };
+                }
+                SimOpenState::FlushRole { mut io, local_role } => {
+                    match Pin::new(&mut io).poll_flush(cx)? {
+                        Poll::Ready(()) => self.state = SimOpenState::ReadRole { io, local_role },
+                        Poll::Pending => {
+                            self.state = SimOpenState::FlushRole { io, local_role };
+                            return Poll::Pending;
+                        }
+                    }
+                }
+                SimOpenState::ReadRole { mut io, local_role } => {
+                    let remote_msg = match Pin::new(&mut io).poll_next(cx)? {
+                        Poll::Ready(Some(msg)) => msg,
+                        Poll::Pending => {
+                            self.state = SimOpenState::ReadRole { io, local_role };
+                            return Poll::Pending;
+                        }
+                        // Treat EOF error as [`NegotiationError::Failed`], not as
+                        // [`NegotiationError::ProtocolError`], allowing dropping or closing an I/O
+                        // stream as a permissible way to "gracefully" fail a negotiation.
+                        Poll::Ready(None) => return Poll::Ready(Err(NegotiationError::Failed)),
+                    };
+
+                    let result = match local_role {
+                        Role::Initiator if remote_msg == Message::Responder => Ok((io, local_role)),
+                        Role::Responder if remote_msg == Message::Initiator => Ok((io, local_role)),
+
+                        _ => Err(ProtocolError::InvalidMessage.into()),
+                    };
```

Comment:
I find matching on a tuple easier to read in such circumstances.

```suggestion
                    let result = match (local_role, remote_msg) {
                        (Role::Initiator, Message::Responder) => Ok((io, local_role)),
                        (Role::Responder, Message::Initiator) => Ok((io, local_role)),
                        _ => Err(ProtocolError::InvalidMessage.into()),
                    };
```